Pull request for 4 sprint

AlexandrSolovyev opened the pull request:
4 Sprint

AlexandrSolovyev committed:
1. Написаны автотесты для проверки main.py
2. Дополнен файл README.md

AlexKitIt commented on main.py:
```diff
@@ -55,3 +55,6 @@ def delete_book_from_favorites(self, name):
     # получаем список Избранных книг
     def get_list_of_favorites_books(self):
         return self.favorites
+
+    def get_books_rating(self):
+        pass
```
Нужно исправить: модуль main изменять нельзя

AlexKitIt commented on tests.py:
```diff
@@ -18,7 +20,65 @@ def test_add_new_book_add_two_books(self):
+        # Тест добавляет одну книгу и проверяет что книга добавилась
+        collector = BooksCollector()
+        collector.add_new_book("Новая книга")
+        assert "Новая книга" in collector.books_genre
```
Нужно исправить: здесь и далее: при наличии методов работы с атрибутом у тестируемой программы, необходимо использовать их, 
нам важно знать что не только атрибут изменился но и метод работы с ним работает, так и тестов будет меньше и атомарность не пострадает

AlexKitIt commented on tests.py:
```diff
@@ -18,7 +20,65 @@ def test_add_new_book_add_two_books(self):
+        collector.delete_book_from_favorites("Book1")
+        assert "Book1" not in collector.favorites
+
+    def test_get_list_of_favorites_books(self):
```
Нужно исправить: уже проверено в  test_add_book_in_favorites

AlexKitIt commented on tests.py:
```diff
@@ -18,7 +20,65 @@ def test_add_new_book_add_two_books(self):
+        collector = BooksCollector()
+        collector.add_new_book(invalid_name)
+        # Проверяем, что книга не была добавлена в словарь books_genre
+        assert invalid_name not in collector.get_books_genre()
```
Нужно исправить: не проверен метод get_books_with_specific_genre

AlexandrSolovyev committed:
1. Переписаны все автотесты для проверки main.py
2. Изменен файл README.md
3. Откатил файл main.py до исходного состояния

AlexKitIt commented on main.py:
```diff
@@ -30,7 +30,7 @@ def get_books_with_specific_genre(self, genre):
 
     # получаем словарь books_genre
-    def get_books_genre(self):
+    def get_books_genre(self) -> object:
```
Повторный коммент: main менять нельзя

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,67 @@
+        assert collector.get_book_genre(book_name) == genre
+
+    @pytest.mark.parametrize("book_name, genre", [("Новая книга", "Фантастика"), ("Старая книга", "Ужасы")])
+    def test_get_book_genre(self, collector, book_name, genre):
```
Нужно исправить: уже проверено в test_set_book_genre не надо копипастить

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,67 @@
-    def test_add_new_book_add_two_books(self):
-        # создаем экземпляр (объект) класса BooksCollector
-        collector = BooksCollector()
+    @pytest.fixture
+    def collector(self):
+        return BooksCollector()
```
Нужно исправить: фикстуры живут в conftest

AlexKitIt commented on tests.py:
```diff
@@ -1,24 +1,67 @@
-    # чтобы тесты были независимыми в каждом из них создавай отдельный экземпляр класса BooksCollector()
\ No newline at end of file
+    @pytest.mark.parametrize("book_name", ["Book1", "Book2"])
+    def test_get_list_of_favorites_books(self, collector, book_name):
```
Нужно исправить: уже проверено в test_add_book_in_favorites не надо копипастить

AlexandrSolovyev committed:
1. Убраны повторяющиеся тесты
2. Изменен файл README.md
3. Откатил файл main.py до исходного состояния
4. Добавлен файл conftests.py с фикстурой
